doc/dot.1
.\" Automatically generated by Pod::Man 4.09 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DOT 1"
.TH DOT 1 "dot-1.2.2" "01/07/2018" "dot manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dot \- manages symbolic links for dotfiles
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdot\fR [\fB\-h\fR|\fB\-\-help\fR][\fB\-c\fR \fIconfigfile\fR] \fIcommand\fR [\fIargs\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This script makes it easy to manage your dotfiles.
.PP
Script is written in bash, and very configurable.
.SH "COMMANDS"
.IX Header "COMMANDS"
.IP "\fBclone\fR" 10
.IX Item "clone"
Clone dotfile repository on your computer with git.
.IP "\fBpull\fR" 10
.IX Item "pull"
Pull the repository from the remote dotfiles repository.
.IP "\fBupdate\fR" 10
.IX Item "update"
Combined command of \fBpull\fR and \fBset\fR commands.
.IP "\fBcd\fR" 10
.IX Item "cd"
Change directory to \fIdotdir\fR.
.IP "\fBlist\fR" 10
.IX Item "list"
Show the list which files will be managed by \fBdot\fR.
.IP "\fBcheck\fR" 10
.IX Item "check"
Check the files are correctly linked to the right places.
.IP "\fBset\fR" 10
.IX Item "set"
Set the symbolic links interactively. This command sets symbolic links configured in \fIdotlink\fR.
.IP "\fBadd\fR" 10
.IX Item "add"
Move the file to the dotfile dir and make its symbolic link to that place.
.IP "\fBedit\fR" 10
.IX Item "edit"
Edit dotlink file \fIdotlink\fR.
.IP "\fBunlink\fR" 10
.IX Item "unlink"
Unlink the selected symbolic links and copy from its original files in the dotfiles repository.
.IP "\fBclear\fR" 10
.IX Item "clear"
Remove the all symbolic links in \fIdotlink\fR.
.IP "\fBconfig\fR" 10
.IX Item "config"
Edit (or create if the file doesn't exist) rcfile \fIdotrc\fR.
.SH "OPTION"
.IX Header "OPTION"
.IP "\fB\-h\fR, \fB\-\-help\fR" 10
.IX Item "-h, --help"
Show the help message.
.IP "\fB\-H\fR, \fB\-\-help\-all\fR" 10
.IX Item "-H, --help-all"
Show man page.
.IP "\fB\-c\fR \fIconfigfile\fR, \fB\-\-config\fR \fIconfigfile\fR" 10
.IX Item "-c configfile, --config configfile"
Specify configuration file to overload. (default: \fI~/.config/dot/dotrc\fR)
.SH "COMMAND OPTIONS"
.IX Header "COMMAND OPTIONS"
.IP "\fBclone\fR [\fIdir\fR]" 4
.IX Item "clone [dir]"
Clone \fIclone_repository\fR onto the specified direction. (default: \fI~/.dotfiles\fR)
.RS 4
.IP "\fB\-f\fR, \fB\-\-force\fR" 6
.IX Item "-f, --force"
Never prompt.
.RE
.RS 4
.RE
.IP "\fBpull\fR [\fB\-\-self\fR]" 4
.IX Item "pull [--self]"
With \-\-self option, update \fBdot\fR itself.
.IP "\fBset\fR" 4
.IX Item "set"
.RS 4
.PD 0
.IP "\fB\-i\fR, \fB\-\-ignore\fR" 6
.IX Item "-i, --ignore"
.PD
No interaction mode (skip all conflicts and do nothing).
.IP "\fB\-f\fR, \fB\-\-force\fR" 6
.IX Item "-f, --force"
Ignore conflicts and force make symbolic links.
.IP "\fB\-b\fR, \fB\-\-backup\fR" 6
.IX Item "-b, --backup"
Ignore conflicts, make symbolic links and create its original backup.
.IP "\fB\-v\fR, \fB\-\-verbose\fR" 6
.IX Item "-v, --verbose"
Print verbose messages.
.RE
.RS 4
.RE
.IP "\fBupdate\fR" 4
.IX Item "update"
Combined command of \fBpull\fR and \fBset\fR commands. Options are same as \fBset\fR commnad ones.
.IP "\fBadd\fR [\fB\-m\fR \fImessage\fR] \fIoriginal_file\fR [\fIdotfile_direction\fR]" 4
.IX Item "add [-m message] original_file [dotfile_direction]"
.RS 4
.PD 0
.IP "\fB\-m\fR \fImessage\fR" 12
.IX Item "-m message"
.PD
Add your comment for dotlink file.
.RE
.RS 4
.RE
.SH "ENVIRONMENT VARIABLE"
.IX Header "ENVIRONMENT VARIABLE"
.IP "\fB\s-1DOT_REPO\s0\fR" 10
.IX Item "DOT_REPO"
Dotfiles repository to manage with this program. You may should change this environment variable. (default: \fIhttps://github.com/ssh0/dotfiles.git\fR)
.IP "\fB\s-1DOT_DIR\s0\fR" 10
.IX Item "DOT_DIR"
Directory for dotfiles (default: \fI\f(CI$HOME\fI/.dotfiles\fR)
.IP "\fB\s-1DOT_LINK\s0\fR" 10
.IX Item "DOT_LINK"
Text file to save symbolic links information (default: \fI\f(CI$DOT_DIR\fI/dotlink\fR)
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
Configuration file is \fB\f(CB$HOME\fB/.config/dot/dotrc\fR and symbolic information is in \fB\f(CB$DOT_DIR\fB/dotlink\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
ssh0 (Shotaro Fujimoto <https://github.com/ssh0>)
.PP
Report bugs here: <https://github.com/ssh0/dot/issues>
.SH "LICENSE"
.IX Header "LICENSE"
This is free software; you may redistribute it and/or modify it under the \s-1MIT\s0 Lisence.


dot.sh
# vim: ft=zsh
# dot - dotfiles management framework

# Version:    1.2.2
# Repository: https://github.com/ssh0/dot
# Author:     ssh0 (Shotaro Fujimoto)
# License:    MIT

DOT_SCRIPT_ROOTDIR="$(builtin cd "$(dirname "${BASH_SOURCE:-${(%):-%N}}")" && pwd)"
readonly DOT_SCRIPT_ROOTDIR
export DOT_SCRIPT_ROOTDIR

dot_main() {

  dot_usage() { #{{{
    cat << EOF
$DOT_COMMAND - Simplest dotfiles manager

Usage: $DOT_COMMAND [options] <commands> [<args>]
  $DOT_COMMAND pull [--self]
  $DOT_COMMAND (set | update) [-i | --ignore] [-f | --force] [-b | --backup] [-v | --verbose]
  $DOT_COMMAND add (<file> [$DOT_DIR/path/to/the/file]) | <symboliclinks>...
  $DOT_COMMAND unlink <symboliclinks>...
  $DOT_COMMAND clear
  $DOT_COMMAND clone [-f | --force] [/dir/to/clone]
  $DOT_COMMAND (-h | --help)

Commands:
  clone   Clone dotfile repository on your computer with git.
  pull    Pull the directory from the remote dotfiles repository.
  cd      Change directory to 'dotdir'.
  list    Show the list which files will be managed by $DOT_COMMAND.
  check   Check the files are correctly linked to the right places.
  set     Set the symbolic links interactively.
  update  Combined command of 'pull' and 'set' commands.
  add     Move the file to the dotfiles directory and make its symbolic link to that place.
  edit    Edit dotlink file.
  unlink  Unlink the selected symbolic links and copy from its original.
  clear   Remove the all symbolic links in 'dotlink'.
  config  Edit (or create if it does not exist) rcfile 'dotrc'.

Options:
  -h, --help      Show this help message.
  -H, --help-all  Show man page.
  -c <file>, --config <file>
                  Specify the configuration file to load.
                  default: \$HOME/.config/dot/dotrc

EOF

  } #}}}

  # Option handling {{{
  local arg
  for arg in "$@"; do
    shift
    case "$arg" in
      "--help") set -- "$@" "-h" ;;
      "--help-all") set -- "$@" "-H" ;;
      "--config") set -- "$@" "-c" ;;
      *)        set -- "$@" "$arg" ;;
    esac
  done

  OPTIND=1
  local dotrc
  while getopts "c:hH" OPT
  do
    case $OPT in
      "c")
        dotrc="$OPTARG"
        ;;
      "h")
        dot_usage
        unset -f dot_usage
        return 0
        ;;
      "H")
        man "${DOT_SCRIPT_ROOTDIR}/doc/dot.1"
        unset -f dot_usage
        return 0
        ;;
      * )
        dot_usage
        unset -f dot_usage
        return 1
        ;;
    esac
  done

  shift $((OPTIND-1))
  # }}}

  # Load common.sh {{{
  source "$DOT_SCRIPT_ROOTDIR/lib/common.sh"
  trap cleanup_namespace EXIT
  # }}}

  # main command handling {{{
  case "$1" in
    clone|pull|update|list|check|set|add|edit|unlink|clear|config|cd)
      subcommand="$1"
      source "$DOT_SCRIPT_ROOTDIR/lib/dot_${subcommand}.sh"
      shift 1
      dot_${subcommand} "$@"
      ;;
    *)
      echo -n "[$(tput bold)$(tput setaf 1)error$(tput sgr0)] "
      echo "command $(tput bold)$1$(tput sgr0) not found."
      return 1
      ;;
  esac

  # }}}

}

eval "alias ${DOT_COMMAND:="dot"}=dot_main"
export DOT_COMMAND

